### PR TITLE
Allow is_mutable to be flipped false via update_metadata_accounts_v2

### PR DESCRIPTION
**Pull request by samuelvanderwaal**
Ported from [metaplex repo:](https://github.com/metaplex-foundation/metaplex/pull/984). 

> Currently there appears to be no way to set is_mutable to be false after an NFT metadata account has been created. Many NFTs, such as any created with default settings by candy_machine_cli, are created with the flag to true. It is desirable to have the ability to flip it to false at a later date, once any metadata issues are resolved or the community makes a decision to do so, etc.

I'll be looking at adding tests to this next before it gets merged. @thlorenz would like to discuss tests with you (it appears the tests for `token-metadata` are currently broken).

### CODE REVIEW

**Comment by thlorenz** [comment]
I will have a look tomorrow afternoon (out until after lunch).

> I'll be looking at adding tests to this next before it gets merged. @thlorenz would like to discuss tests with you (it appears the tests for token-metadata are currently broken).

There are currently no token-metadata tests on master (I'm working on them on a separate branch).
What you maybe referring to is a `main` script under the `tests` folder which currently doesn't build as it is mainly here to inform writing integration tests driven by the contract's JS SDK.

However for your PR all checks passed so you're good for now.

Tests aren't needed until I finished the [amman](https://github.com/metaplex-foundation/amman) test utils that will help you with that. Also I'll merge in the tests I added to token-metadata so far to serve as a guidance for you.

**Comment by thlorenz** [comment]
Just realizing you may not be aware that there are Rust tests for token-metadata already which you see passing [here](https://github.com/metaplex-foundation/metaplex-program-library/runs/4464698849?check_suite_focus=true).
They are defined [here](https://github.com/metaplex-foundation/metaplex-program-library/tree/master/token-metadata/program/tests) and you should add tests for your feature add here as well.

However as mentioned integration tests will have to come a bit later.

**Comment by samuelvanderwaal** [comment]
> Just realizing you may not be aware that there are Rust tests for token-metadata already which you see passing [here](https://github.com/metaplex-foundation/metaplex-program-library/runs/4464698849?check_suite_focus=true). They are defined [here](https://github.com/metaplex-foundation/metaplex-program-library/tree/master/token-metadata/program/tests) and you should add tests for your feature add here as well.
> 
> However as mentioned integration tests will have to come a bit later.

Yeah I was looking at the `test` folder at the outer level with `main.rs` that doesn't build.

Not trying to rush you, just wanted to make sure I knew what I should be doing for tests. I'll add Rust tests for my new update function tomorrow and then we can add integration tests once you've got the framework all complete.

Thanks!

**Review by thlorenz — APPROVED** [review]
Awesome, this is looking good.
I like that you added tests for the invalid input cases as well.

Also thanks for fixing the `test/main.rs` as part of this.

Do we need to update the SDK next to be able to call the new instruction?
If so I should have the testing foundation by EOW so you could do this with an integration test included .. all in a separate PR though :)

**Comment by samuelvanderwaal** [comment]
> Awesome, this is looking good. I like that you added tests for the invalid input cases as well.
> 
> Also thanks for fixing the `test/main.rs` as part of this.
> 
> Do we need to update the SDK next to be able to call the new instruction? If so I should have the testing foundation by EOW so you could do this with an integration test included .. all in a separate PR though :)

Yeah, I can look at doing that next week after your testing foundation is ready.

I'll let Austin review this PR and then will merge it if he doesn't find any blockers.

**Comment by novaknole** [comment]
Hi folks,

This feature is extra handy. Currently, Updating `isMutable` from js library is not possible. But I went ahead and quickly add it, though, tx immediatelly gets failed with `BoschError` and couldn't detect why.

Is this contract already updated on metaplex for devnet and mainnet as well ? I tested it for devnet.

Would appreciate your inputs on this.

**Comment by samuelvanderwaal** [comment]
> Hi folks,
> 
> This feature is extra handy. Currently, Updating `isMutable` from js library is not possible. But I went ahead and quickly add it, though, tx immediatelly gets failed with `BoschError` and couldn't detect why.
> 
> Is this contract already updated on metaplex for devnet and mainnet as well ? I tested it for devnet.
> 
> Would appreciate your inputs on this.

This fix should be deployed on devnet currently and will be deployed onto mainnet likely next week.

**Comment by samuelvanderwaal** [comment]
> @samuelvanderwaal When was it deployed on `devnet` ? I created candy machine 2 days ago with devnet and passing `isMutable:0` still fails with `BoschError`.

I just tried this on devnet and it worked. How are you calling the handler?

**Comment by truedandyman0617** [comment]
> > @samuelvanderwaal When was it deployed on `devnet` ? I created candy machine 2 days ago with devnet and passing `isMutable:0` still fails with `BoschError`.
> 
> It's definitely deployed to testnet because I just called the handler and it worked. The JS library is probably not updated yet, though.

If it has deployed on devnet already, perhaps it looks like made another issue. When I try to update metadata and set to isMutable: true without any extra setting, I get the error "Is Mutable can only be flipped to false".
And when I try to set this value to false in CLI code, it is updated without error but it sets to immutable.
I am not sure if it's not bug but in this case we can only update it one time.

**Comment by samuelvanderwaal** [comment]
> > > @samuelvanderwaal When was it deployed on `devnet` ? I created candy machine 2 days ago with devnet and passing `isMutable:0` still fails with `BoschError`.
> > 
> > 
> > It's definitely deployed to testnet because I just called the handler and it worked. The JS library is probably not updated yet, though.
> 
> If it has deployed on devnet already, perhaps it looks like made another issue. When I try to update metadata and set to isMutable: true without any extra setting, I get the error "Is Mutable can only be flipped to false". And when I try to set this value to false in CLI code, it is updated without error but it sets to immutable. I am not sure if it's not bug but in this case we can only update it one time.

This is how it's supposed to work. You can't set `isMutable: true`. If it's already true, you don't need to set it, and if it's false, you can't set it to be mutable again. (Can only be flipped false.) You can set it to false but not back to mutable again. 

Let me know if that makes sense.

**Comment by samuelvanderwaal** [comment]
The idea here is that an NFT being mutable: false (immutable) is a guarantee that the NFT can't be changed any more. Once you set an NFT to mutable: false, that's it, it can't be updated anymore and the owner has a guarantee that no one can mess with their NFT. If we allowed is mutable to be set back and forth there would be no point in having that field at all.

**Comment by truedandyman0617** [comment]
When I use CLI, I don't set that flag but the updateMetadata function set it to true always in [Here](https://github.com/metaplex-foundation/metaplex/blob/master/js/packages/cli/src/commands/mint-nft.ts#L250).
And we should be able to set it to true anytime if current value is true.
How this could be an error?

**Comment by samuelvanderwaal** [comment]
> When I use CLI, I don't set that flag but the updateMetadata function set it to true always in [Here](https://github.com/metaplex-foundation/metaplex/blob/master/js/packages/cli/src/commands/mint-nft.ts#L250). And we should be able to set it to true anytime if current value is true. How this could be an error?

I see, you're using the JS library. Can you open an issue on that repo describing the bug? This PR was for updating the *on-chain* Solana program.

**Comment by truedandyman0617** [comment]
Well, I think current js library is ok since isMutable is optional and it can has the value True as default.
The problem is, when I try update Metadata for the mutable nft, the TokenMetadata program bans it with the error "Is Mutable can only be flipped to false". It looks like we can update Metadata only 1 time, after that NFT should be immutable always.
This is the problem now and that's why I opened issue in here.

**Comment by samuelvanderwaal** [comment]
In the on-chain Solana program, `is_mutable` is an `Option` in the Rust code. This means you can pass in either a `None` value or `Some(true)` or `Some(false)`. If you pass in `None`, nothing happens and the value stays the same. If you pass in `Some(true)` you'll get an error because you're saying you want to change it to be `true` and you can't change it to be `true` because it can only be flipped to `false`. If you pass in `Some(false)` that will set it to `false` (or keep it that way if it's already `false`).

Unless you're trying to change the `is_mutable` value you should always pass in a `None`. Typically this is a `null` value in JS libraries. If our JS libraries do not allow passing in a `null` then that's a bug in the JS libraries, but as far as I can tell the on-chain program is still working as expected. 

If you can show me a specific test you're running where you think the on-chain program is not behaving correctly, that would be helpful. Otherwise, if this is an issue specifically with our JS library you can open an issue directly there.

@aheckmann pinging you for context because I think this is about how the JS library handles the `updateMetadataAccountsV2` command.

**Comment by truedandyman0617** [comment]
> In the on-chain Solana program, `is_mutable` is an `Option` in the Rust code. This means you can pass in either a `None` value or `Some(true)` or `Some(false)`. If you pass in `None`, nothing happens and the value stays the same. If you pass in `Some(true)` you'll get an error because you're saying you want to change it to be `true` and you can't change it to be `true` because it can only be flipped to `false`. If you pass in `Some(false)` that will set it to `false` (or keep it that way if it's already `false`).
> 
> Unless you're trying to change the `is_mutable` value you should always pass in a `None`. Typically this is a `null` value in JS libraries. If our JS libraries do not allow passing in a `null` then that's a bug in the JS libraries, but as far as I can tell the on-chain program is still working as expected.
> 
> If you can show me a specific test you're running where you think the on-chain program is not behaving correctly, that would be helpful. Otherwise, if this is an issue specifically with our JS library you can open an issue directly there.
> 
> @aheckmann pinging you for context because I think this is about how the JS library handles the `updateMetadataAccountsV2` command.

I understood exactly, Yes it looks like JS library code issue. Thanks.

**Pull request merged**
